### PR TITLE
win_pkg New Reporting for genrepo refresh_db & Fix #35342 #34781, New Unicode Support Software Names

### LINKED ISSUE #34781: winrepo-ng jinjia2 compilation error is too quiet - needs "displaying"
see new comment below here https://github.com/saltstack/salt/issues/34781#issuecomment-233854043, this turned out to be another case where a jinjja2 template compilation simply failed silently with no info whatsoever being displayed.: 

### Description of Issue/Question
@twangboy help I just cannot get it to use some newly added winrepo-ng pkg files to install them. Pretty sure I am not doing anything wrong or different from previous versions.
 
### Setup
I create a new winrepo-ng file on the saltstack github site (as usual), then I git update my master and refresh_db my windows minions, but ```pkg.install my-brand-new-pkg```, reports that said pkg cannot be located. Although I can see the file being copied to the cache during the refresh_db cmd. but a pgk.get_repo_data does indeed not show me the newly added pkg.

### Steps to Reproduce Issue
```salt my-win-minion pkg.refresh_db```
```salt my-win-minion pkg.install check_mk-agent-msi```
and it cannot locate the check_mk-agent-msi file. 
(Note: there is a number of check-mk-agent* pkgs, ```'check-mk'``` not ```'check_mk'``` they are 'older' and I need to test some new ones under the name of 'check_mk-agent-msi')

### Versions Report
both master and minion are running on 2016.3.1

### LINKED ISSUE #35342: win_pkg: refresh_db doesn't remove cached items which have been renamed or removed
If you remove or rename items in the repo-ng folder on the master, these are not removed from the minion cache when you run win_pkg.refresh_db. Particularly in the case of renaming, it's very likely to generate a name collision for the package which may result in incorrect installation states being returned (the pkg name map can still reference old sls files).

As an example, I'll rename my 7zip directory. It's already cached under it's old name:

<pre>salt salt-test cmd.run 'dir /b c:\salt\var\cache\salt\minion\files\base\win\repo-ng\*zip' 
salt-test:
    7zip</pre>

I rename it:

<pre>mv /srv/salt/win/repo-ng/7zip /srv/salt/win/repo-ng/seven_zip</pre>

Then run win_pkg.refresh_db, checking that it only lists the renamed directory:

<pre>salt salt-test pkg.refresh_db | grep zip
    - c:\salt\var\cache\salt\minion\files\base\win\repo-ng\seven_zip\7zip.sls</pre>

But on the minion, I've now got two folders in the cache:

<pre>salt salt-test cmd.run 'dir /b c:\salt\var\cache\salt\minion\files\base\win\repo-ng\*zip' 
salt-test:
    7zip
    seven_zip</pre>

And worse still the sls file is still there, and this can be matched when generating the name map:

<pre>salt salt-test cmd.run 'type c:\salt\var\cache\salt\minion\files\base\win\repo-ng\7zip\7zip.sls' 
salt-test:
    {% if grains['cpuarch'] == 'AMD64' %}
      {% set name_arch = ' (x64 edition)' %}
      {% set installer_arch = '-x64' %}
    {% else %}
      {% set name_arch = '' %}
      {% set installer_arch = '' %}
    {% endif %}
    
    7zip:
      '15.14.00.0':
        full_name: '7-Zip 15.14{{ name_arch }}'
        installer: 'salt://win/repo-ng/7zip/7z1514{{ installer_arch }}.msi'
        install_flags: '/qn /norestart'
        uninstaller: '{23170F69-40C1-2702-1514-000001000000}'
        uninstall_flags: '/qn /norestart'
        refresh: True
        msiexec: True
        use_scheduler: False
        locale: en_US
        reboot: False
        cache_dir: False
      '9.20.00.0':
        full_name: '7-Zip 9.20{{ name_arch }}'
        installer: 'salt://win/repo-ng/7zip/7z920{{ installer_arch }}.msi'
        install_flags: '/qn /norestart'
        uninstaller: '{23170F69-40C1-2702-0920-000001000000}'
        uninstall_flags: '/qn /norestart'
        refresh: True
        msiexec: True 
        use_scheduler: False
        locale: en_US
        reboot: False
        cache_dir: False</pre>

I can't think of any situation where you would want to maintain items in the repo-ng cache when they don't exist on the master, and trying to clear the cache with saltutil is going to clear the entire cache so I imagine this shouldn't be used during a state run.

So that the master side can be edited, surely pkg.refresh_db needs to be removing anything that doesn't exist any longer?

Tested on 2016.3.2

### PR DESCRIPTION
**Pull request by damon-atkins**
### What does this PR do?

Fix #35342
 Issue: win_pkg: refresh_db doesn't remove cached items which have been
renamed or removed
 Solution: Conservative solution was to just remove *.sls files from
\win\repo-ng  tree  as part of pkg.refresh_db instead of removing all
of \win\repo-ng\salt-winrepo-ng

Fix saltstack/salt#34781
 Issue: Logs did not contain enough information, and from executing it from the
master it was too difficult to determine if there was an issue due to the
output being data
 Solution:
- pkg.genrepo now reports the status of the operation instead of the
  large about of metadata. The status report indicates:
- Total SLS Files processed
- Number of failed sls files
- Number of successful sls files
- list of files with errors and the render/compile errors they had
- optionally a list of successfully render/compiled files and the
  package names within those files
- pkg.refresh_db was also part of this issue, as it just reported a list
  of files copied from the same large list of files would be reported almost
  every time. As pkg.refresh_db calls pkg.genrepo after copy all the files
  it now reports the same as data as pkg.genrepo.

Unicode Support
  Issue: Windows Registry module supports Unicode in development, win_pkg
needed to be updated to support the Unicode output from reg.
### Previous Behavior

pkg.genrepo return the data structure of the sls files, to did not report few issues to local logs, did not report any issues back to the master.

pkg.refresh_db
*provide a list of files copy from the master, the list of files copied was long always, as it was not a sync process.
- called pkg.genrepo but did not report any issues or errors.
### New Behavior

Both pkg.genrepo and pkg.refresh_db report the same output from pkg.genrepo.
pkg.genrepo reports the status of render the files.

pkg.list_pkgs
Supports Unicode. It was dependant on Unicode changes to reg.py.
See Windows Registry Unicode Support and PY2 character encoding (module.reg) saltstack/salt#32835
### Tests written?

No Extra Tests.   Tested against salt-winrepo-ng sls files.

### CODE REVIEW

**Comment by damon-atkins** [comment]
Seems these failures are not part of this PR. **Please merge into Develop.**
```
unit.states.dockerng_test.DockerngTestCase.test_image_present_and_force
unit.states.grains_test.GrainsTestCase.test_present_fail_overwrite
unit.states.grains_test.GrainsTestCase.test_present_overwrite
integration.states.file.FileTest.test_managed_file_mode_keep_local_source
integration.states.file.FileTest.test_template_local_file
integration.states.file.FileTest.test_template_local_file_noclobber
```

**Comment by damon-atkins** [comment]
Not sure why jenkins/PR/salt-pr-linode-ubuntu14-n  is reporting
```
93153 blocks
++ git rev-parse HEAD
+ '[' 1bb27255cb81d2ce7f0705418eb77e28e439358a '!=' 1cd499cb7fa5c2ca068613ecfc20d5881cf4ff87 ']'
++ git rev-parse HEAD
+ echo 'Commits don'\''t match 1bb27255cb81d2ce7f0705418eb77e28e439358a == 1cd499cb7fa5c2ca068613ecfc20d5881cf4ff87'
Commits don't match 1bb27255cb81d2ce7f0705418eb77e28e439358a == 1cd499cb7fa5c2ca068613ecfc20d5881cf4ff87
+ echo 'Stopping execution'
```

**Comment by damon-atkins** [comment]
Thanks @meggiebot look forward to the feedback.
Interesting seems to of grabbed someone elses PR
https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/4875/console
```
16:32:22 ************* Module salt.modules.mongodb
16:32:22 salt/modules/mongodb.py:488: [E8201(whitespace-after-left-parenthesis), ] PEP8 E201: whitespace after '('
16:32:22 salt/modules/mongodb.py:488: [E8202(whitespace-before-right-parenthesis), ] PEP8 E202: whitespace before ')'
16:32:22 salt/modules/mongodb.py:473: [C0303(trailing-whitespace), ] Trailing whitespace
16:32:22 salt/modules/mongodb.py:484: [C0303(trailing-whitespace), ] Trailing whitespace
16:32:22 salt/modules/mongodb.py:488: [C0326(bad-whitespace), ] No space allowed after bracket
16:32:22             ids = col.update_one( _id_field, {'$set': _update_doc} )
16:32:22                                 ^
16:32:22 salt/modules/mongodb.py:488: [C0326(bad-whitespace), ] No space allowed before bracket
16:32:22             ids = col.update_one( _id_field, {'$set': _update_doc} )
16:32:22                                                                    ^
16:32:22 salt/modules/mongodb.py:492: [C0303(trailing-whitespace), ] Trailing whitespace
```

**Comment by damon-atkins** [comment]
Remove a blank line to pull the focus back on this PR

**Comment by damon-atkins** [comment]
Hi @rallytime seem to be getting lint issues from salt/modules/mongodb.py in develop outside of this PR.

**Comment by rallytime** [comment]
Hi @damon-atkins - Yeah I just fixed those at the HEAD of the develop branch. Don't worry about those for now. :)

**Comment by rallytime** [comment]
Thank you for your work here @damon-atkins!

File: salt/modules/win_pkg.py [hunk 1]
```diff
@@ -402,63 +482,208 @@ def _get_local_repo_dir(saltenv='base'):
...
     '''
-    Generate winrepo_cachefile based on sls files in the winrepo
+    Returns age in seconds of the generated repo
+    '''
+    (repo_remote, repo_local) = _get_repo_src_dest(saltenv)
+    if not os.path.exists(repo_local):
+        os.makedirs(repo_local)
+    winrepo = os.path.join(repo_local, 'winrepo.p')
+
+    try:
```
Comment (terminalmage) [inline salt/modules/win_pkg.py]: What is the point of a try/except here, if all we do is raise the exception? :confused:

**Comment by damon-atkins** [comment]
I plan on doing another review of the code for example 'winrepo.p' is hard coded, and has been since day 1 by the look of it,  but I notice in the doco their is a setting to set the name of for  'winrepo.p'  

Re your commends it should most likely just return -1 rather than raise an issue.   Was not sure how to capture permission problems on windows.  Going to put more smarts into _get_repo_src_dest  pull some of the repeated code out in other sections.

**Comment by damon-atkins** [comment]
@meggiebot some items for release notes:
- Fix: win_pkg old sls files used to descibe software packages  were not being removed on the minions
- New: Improved Unicode Support
- Removal of  win_pkg `available_version()` please use `latest_version()` instead.
See also PR #36222

**Pull request merged**